### ISSUE TITLE
Back reference feature.

### ISSUE THREAD

**Issue by ldmi3i** [j-easy/easy-random#444]
I suggest a feature - provide support of back references. 
To provide correct graph for ORM frameworks when using of bidirectional relation between tables.
For example if Root class have 
`List<Child> children;` 
 and Child have field
`Root root;`
populate root field with correct reference of existing object instead of generating new.
And the same for Many to Many relationship.

The function can be roughly expressed by a set of integration tests (class definitions under tests)
```java
    @Test
    void backReferenceOneToManyPopulate_correctPopulated() {
        EasyRandomParameters easyRandomParameters = new EasyRandomParameters().enableBackReference();
        EasyRandom easyRandom = new EasyRandom(easyRandomParameters);

        RootOM generatedRootOM = easyRandom.nextObject(RootOM.class);

        //check that all first line children has correct parent
        assertTrue(generatedRootOM.getFirstChildList().stream()
                .allMatch(firstChild -> (firstChild.getRootClass() != null) && (firstChild.getRootClass() == generatedRootOM)));
        assertTrue(generatedRootOM.getSecondChildList().stream()
                .allMatch(secondChild -> (secondChild.getRootClass() != null) && (secondChild.getRootClass() == generatedRootOM)));

        //full list of link objects
        List<ChildLinkOM> childLinkOMList = Stream.concat(
                generatedRootOM.getFirstChildList().stream()
                        .flatMap(firstChild -> firstChild.getChildLinkList().stream()),
                generatedRootOM.getSecondChildList().stream()
                        .flatMap(secondChild -> secondChild.getChildLinkList().stream())
        ).distinct().collect(Collectors.toList());

        //check full data consistency
        assertTrue(childLinkOMList.stream()
                .allMatch(childLinkOM ->
                        (childLinkOM.getFirstChild().getRootClass() == childLinkOM.getSecondChild().getRootClass())
                ));
    }

    @Test
    void backReferenceMultiRootOneToManyPopulate_correctPopulated() {
        EasyRandomParameters easyRandomParameters = new EasyRandomParameters().enableBackReference();
        EasyRandom easyRandom = new EasyRandom(easyRandomParameters);

        List<RootOM> rootObjects = easyRandom.objects(RootOM.class, 10).collect(Collectors.toList());

        //check that all first line children has correct parent
        assertTrue(rootObjects.stream().allMatch(root -> root.getFirstChildList().stream()
                .allMatch(firstChild -> (firstChild.getRootClass() != null) && (firstChild.getRootClass() == root))));
        assertTrue(rootObjects.stream().allMatch(root ->root.getSecondChildList().stream()
                .allMatch(secondChild -> (secondChild.getRootClass() != null) && (secondChild.getRootClass() == root))));

        //full list of link objects
        List<ChildLinkOM> childLinkOMList = Stream.concat(
                rootObjects.stream().flatMap(root -> root.getFirstChildList().stream()
                        .flatMap(firstChild -> firstChild.getChildLinkList().stream())),
                rootObjects.stream().flatMap(root -> root.getSecondChildList().stream()
                        .flatMap(secondChild -> secondChild.getChildLinkList().stream()))
        ).distinct().collect(Collectors.toList());

        //check full data consistency
        assertTrue(childLinkOMList.stream()
                .allMatch(childLinkOM ->
                        (childLinkOM.getFirstChild().getRootClass() == childLinkOM.getSecondChild().getRootClass())
                ));
    }

    @Test
    void backReferenceManyToManyPopulate_correctPopulated() {
        EasyRandomParameters easyRandomParameters = new EasyRandomParameters().enableBackReference();
        EasyRandom easyRandom = new EasyRandom(easyRandomParameters);

        RootMM root = easyRandom.nextObject(RootMM.class);

        assertTrue(root.getFirstChildMM().stream()
                .allMatch(firstChildMM ->
                        firstChildMM.getRootMMList().contains(root) &&
                                firstChildMM.getRootMMList().size() == 1
                ));
        assertTrue(root.getSecondChildMM().stream()
                .allMatch(secondChildMM ->
                        secondChildMM.getRootMMList().contains(root) &&
                                secondChildMM.getRootMMList().size() == 1
                ));

        //full list of link objects
        List<ChildLinkMM> childLinkMMList = Stream.concat(
                root.getFirstChildMM().stream()
                        .flatMap(firstChild -> firstChild.getChildLinkMMSet().stream()),
                root.getSecondChildMM().stream()
                        .flatMap(secondChild -> secondChild.getChildLinkMMSet().stream())
        ).distinct().collect(Collectors.toList());

        //check full data consistency
        childLinkMMList
                .forEach(childLinkMM -> {
                            List<RootMM> rootThroughFirst = childLinkMM.getFirstChildMMList().stream()
                                    .flatMap(firstChildMM -> firstChildMM.getRootMMList().stream())
                                    .distinct()
                                    .sorted(Comparator.comparing(RootMM::getId))
                                    .collect(Collectors.toList());
                            List<RootMM> rootThroughSecond = childLinkMM.getSecondChildMMList().stream()
                                    .flatMap(secondChildMM -> secondChildMM.getRootMMList().stream())
                                    .distinct()
                                    .sorted(Comparator.comparing(RootMM::getId))
                                    .collect(Collectors.toList());
                            assertIterableEquals(rootThroughFirst, rootThroughSecond);
                        }

                );
    }
```
Class structure something like following with getters and setters
```java
public class RootOM {
    private List<FirstChildOM> firstChildOMList;
    private List<SecondChildOM> secondChildOMList;
}

public class SecondChildOM {
    private RootOM rootOM;
    private List<ChildLinkOM> childLinkOMList;
}

public class FirstChildOM {
    private RootOM rootOM;
    private List<ChildLinkOM> childLinkOMList;
}

public class ChildLinkOM {
    private FirstChildOM firstChildOM;
    private SecondChildOM secondChildOM;
}
```
For many to many it same but with lists everywhere.
For details you can look to my fork where this tests written. And some drafts there as well.

**Comment by benas** [comment]
Thank you for suggesting this feature. However, the project is in [maintenance mode](https://github.com/j-easy/easy-random#project-status) now, so we are not going to add any new features anymore. If this feature cannot be achieved using custom randomizers and/or a custom `ObjectFactory`, then it can be implemented in a fork.

**Comment by JohannesBe** [comment]
If there is ever anyone who would like this for generating JPA objects, the following code more or less achieves this:

```
public class BiDirectionalObjectFactory extends ObjenesisObjectFactory {
    @Override
    public <T> T createInstance(Class<T> type, RandomizerContext context) {
        T instance = super.createInstance(type, context);

        // When the root object is initialized the context is (still as good as) empty
        if (context.getCurrentObject() == null) return instance;

        Field field = ReflectUtils.findFieldInClassOrSuper(
                context.getCurrentObject().getClass(), context.getCurrentField()
        );

        if (field.isAnnotationPresent(OneToOne.class)) {
            OneToOne o2o = field.getAnnotation(OneToOne.class);

            if (!o2o.mappedBy().equals("")) {
                trySetting(
                        ReflectUtils.findFieldInClassOrSuper(type, o2o.mappedBy()),
                        instance,
                        context.getCurrentObject()
                );
            }
        }
        else if (field.isAnnotationPresent(OneToMany.class)) {
            OneToMany o2m = field.getAnnotation(OneToMany.class);
            if (!o2m.mappedBy().equals("")) {
                trySetting(
                        ReflectUtils.findFieldInClassOrSuper(type, o2m.mappedBy()),
                        instance,
                        context.getCurrentObject()
                );
            }
        }

        return instance;
    }
}
```

And then when you create your parameters, don't forget to:

```
new EasyRandomParameters()
    // ...
    .objectFactory(new BiDirectionalObjectFactory());
```

**Comment by JohannesBe** [comment]
@benas using a similar methodology to the above it might be interesting to add a `@BiDirectional(references="field")` annotation, or alternately extend the Randomizer annotation to include a backReference property `@Randomizer(backReference="field")` . I don't think it would require a lot of effort and it might add a lot of value.

Or is maintenance mode a hard no-no for any non-bugs for you?

**Comment by ldmi3i** [comment]
@JohannesBe you proposed nice solution. I tested your factory a little. It works good when there is all O:M connections: Parent : O:M : Child : O:M : ChildChild .
But in case like: Parent: O:M : Child : O:M : ChildChild : M:O : OtherChild it breaks (many root elements are created).
And. In case of more than 1 level depth
this
```
        Field field = ReflectUtils.findFieldInClassOrSuper(
                context.getCurrentObject().getClass(), context.getCurrentField()
        );
```
needs to be replaced with the following
```
            Field field = findFieldInClassOrSuper(
                    context.getCurrentObject().getClass(), context.getCurrentField().substring(context.getCurrentField().lastIndexOf(".") + 1)
            );
```
because of field name concatenating.

I think it can be complex to process JPA entities correctly even with O:O and O:M relations (because there is many cases). And this may require further elaboration of this direction, which, as I understand it, is not assumed by the author.

In general, I think it can be very useful to prepare the random data for load testing or any other tests, so as not to write huge logic to populate many correlated data.

**Comment by JohannesBe** [comment]
I've investigated the case where the back-references turns invalid, and indeed it seems that when nesting deeply this occurs. I think this is because at that point the `ojectPool` becomes saturated. I tried setting the parameters `objectPoolSize()` to `Integer.MAX_VALUE`, but apparently that value is used to initialize an array somewhere, hence pre-allocating way too much space and throwing OOM's. Setting it to 0 (e.g. to disable it) is also not possible it seems, so this appears to be a dead-ish end.
